app/services/authorizer.rb

```diff
@@ -1,19 +1,23 @@
 class Authorizer
+  include AuthorizerCache
+
   attr_accessor :user, :base_collection, :organization_ids, :location_ids
 
   def initialize(user, options = {})
-    @cache = HashWithIndifferentAccess.new { |h, k| h[k] = HashWithIndifferentAccess.new }
+    initialize_cache
     self.user = user
     self.base_collection = options.delete(:collection)
   end
 
-  def can?(permission, subject = nil)
+  def can?(permission, subject = nil, cache = true)
     if subject.nil?
       user.permissions.where(:name => permission).present?
     else
       return true if user.admin?
-      collection = @cache[subject.class.to_s][permission] ||= find_collection(subject.class, :permission => permission)
-      collection.include?(subject)
+      return collection_cache_lookup(subject, permission) if cache
+
+      find_collection(subject.class, :permission => permission).
+        where(:id => subject.id).present?
```

Comment:
Prefer `any?` to `present?` when dealing with presence and AR collections, it optimises it using a COUNT so doesn't need to retrieve the contents of the columns.  This currently does:

    2016-07-12T09:43:44 [sql] [D]   Host::Managed Load (0.2ms)  SELECT "hosts".* FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("hosts"."name" = 'wstest.example.com')) AND "hosts"."id" = ?  [["id", 118]]

but with #any? it does:

    2016-07-12T09:49:12 [sql] [D]    (0.2ms)  SELECT COUNT(*) FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("hosts"."name" = 'wstest.example.com')) AND "hosts"."id" = ?  [["id", 118]]